docs/mailcivi.man
.TH man 1 "September 16, 2014"
.LO 1
.SH NAME
mailcivi \- read mailing body text and create a mail template from it in CiviCRM
.SH SYNOPSIS
.B man
.RB [ \-vh ]
.RB [ \--civicrm
.IR url ]
.RB [ \--action
.RB {
.IR nothing
.R |
.IR create
.R |
.IR send
.RI }
.IB ]
.RB [ \--sitekey
.IR key ]
.RB [ \--apikey
.IR key ]
.RB [ \--name
.IR name ]
.RB [ \--subject
.IR subject ]
.RB [ \--from_id
.IR id ]
.RB [ \--to_id
.IR id ]
.RB [ \--to_name
.IR groupname ]
.RB [ \--json
.IR file ]
.RB [ \--url
.IR url ]
.RB [ \--html
.IR file ]
.RB [ \--text
.IR file ]
.RB [ \--text
.IR file ]

.SH DESCRIPTION
.B mailcivi
is a script to read text forming the proposed body of an email and create a MailTemplate
object in a
.I CiviCRM
instance. The body text can be provided in several ways: from local
files, from a local JSON array, or from a remote URL supplying JSON data.
.PP
Once the MailTemplate object is created, the mail actually can be further modified and/or
sent using the normal
.I CiviCRM
user interface.

.SH OPTIONS
.TP
.B \-\-\^civicrm " url"
Specify the URL for the remote CiviCRM instance. The URL must include the path to the
civicrm module, and not just the homepage. For example in Drupal/Civi:
.BR http://example.com/sites/all/modules/civicrm .
.TP
.B \-\-action " action"
What to do:
.RS
.IP \(bu 4
"nothing" \- do not create a template or send it.
.IP \(bu 4
"create" \- create a template on CiviCRM but do not send it.
.IP \(bu 4
"send" \- create a template on CiviCRM and also send it, if the recipient group has also been defined.
.RE
.TP
.B \-\-sitekey " key"
The CiviCRM site key, which is part of the installation configuration.
.TP
.B \-\-apikey " key"
The user's API key, which is part of the user profile. There is no user-interface within
.I CiviCRM
to create an api key.
.TP
.BR \-\-name " name "
A string which will be used to name the mail template within
.I CiviCRM
but which is not visible to email recipients.
.TP
.BR \-\-subject " subject "
A string which will be used as the proposed subject line in the new mail template.
.TP
.BR \-\-from_id " id "
A small integer identifying the CiviCRM user which will own the mail template and which
by default will be named as the sender of any email.
.TP
.BR \-\-to_id " id "
A small integer identifying the CiviCRM group ID to which this template will be addressed.
Only used if 'action' is 'send'. Only one of to_id or to_name can be set.
.TP
.BR \-\-to_name " name "
A small integer identifying the CiviCRM group name to which this template will be addressed.
Only used if 'action' is 'send'. Only one of to_id or to_name can be set.
.TP
.BR \-\-json " file"
A file containing JSON-encoded data in four fields:
.RS
.IP \(bu 4
"subject" \- as for the --subject command line option.
.IP \(bu 4
"from_id" \- as for the --from_id command line option.
.IP \(bu 4
"name" \- as for the --name command line option.
.IP \(bu 4
"html" \- a file containing the HTML version of the body of the email template.
.IP \(bu 4
"plain" \- optional file containing the plain text version of the body of the email template. If not present, it
will be created from the HTML version.
.RE

.SH JSON DATA
If used with either the file or url sources, the JSON data file must have the following form.
.nf
.RS
{
  "name": "The name of the template in Civi",
  "action": "create",
  "subject": "The mail template subject line",
  "creator_id": "960",
  "from_email": "staff@example.org",
  "from_name": "example",
  "recipient": "",
  "html": ".. html text ..",
  "plain": ".. plain text .."
}
.RE
.fi

.SH EXAMPLES
.PP
The following reads a JSON data source at the example.com domain and constructs a template on
the server at the example.org domain. No --action is specified, so the template is only created
and no schedule is set for sending it.
.PP
.nf
.RS
mailcivi --civicrm=http://example.org/sites/all/modules/civicrm \\
   --url=http://example.com/content_json/vor \\
   --sitekey=fddfuhef1153ac --apikey=34o8vbeofu
.RE
.fi

.PP
The following reads the template data a local JSON file and creates a template on the server
at the example.org domain. The recipients are set to those in the group with id '31' and assuming
that such a group exists, the mail will be sent immediately.
.PP
.nf
.RS
mailcivi --action=send \\
   --civicrm=http://example.org/sites/all/modules/civicrm \\
   --json=local.json \\
   --to_id=31 \\
   --sitekey=fddfuhef1153ac --apikey=34o8vbeofu
.RE

.fi
.PP
The following reads the template data a local JSON file and creates a template on the server
at the example.org domain. The recipients are set to those in the group with id '31' and assuming
that such a group exists, the mail will be sent immediately.

.SH SEE ALSO
.IP \(bu 4
python-civicrm \- project on Github that wraps the CiviCRM API as a Python object.

.SH AUTHOR
.B mailcivi
and this manual page was written by Ruth Ivimey-Cook for eLifesciences.


mailcivi.py
#!bin/python
# Script to read an email template, either directly or as a JSON bundle,
# and send that bundle to a CiviCRM instance running remotely. JSON bundle
# data can be read from a local file or from a web service via a supplied
# URL.
#
# Written: August 2014 Ruth Ivimey-Cook
#

from __future__ import print_function
import sys
import os
import argparse
import json
import html2text
import requests

basedir = os.path.dirname(os.path.abspath(__file__))
os.sys.path.insert(0, os.path.join(basedir, 'pythonCivicrm'))

from pythoncivicrm.pythoncivicrm import ( CiviCRM, CivicrmError )


def debugmsg(settings, *objs):
    """
    Print additional info to stderr iff the verbose flag has been enabled
    """
    if settings.verbose > 1:
        print("DEBUG: ", *objs, end='\n', file=sys.stderr)


def infomsg(settings, *objs):
    """
    Print additional info to stderr iff the verbose flag has been enabled
    """
    if settings.verbose:
        print("INFO: ", *objs, end='\n', file=sys.stderr)


def warningmsg(settings, *objs):
    """
    Print warning message to stderr
    """
    print("WARNING: ", *objs, end='\n', file=sys.stderr)


def errormsg(settings, *objs):
    """
    Print error message to stderr
    """
    print("ERROR: ", *objs, end='\n', file=sys.stderr)


class CiviMailTemplate:
    """
    Mail template object, filled out by readjson() and readlocal(),
    to hold the intermediate form of the input email.
    """
    pass


def getoptions(settings):
    """
    Use the Python argparse module to read in the command line args
    """
    parser = argparse.ArgumentParser(
        prog='mailcivi',
        description='Create a new mail template in a remote CiviCRM installation.',
        usage='%(prog)s [options] (--json=file|--html=file) [--text=file]'
    )
    parser.add_argument('--verbose', '-v',
                        action='count',
                        help='Print additional messages to stderr',
                        default=0)
    parser.add_argument('--civicrm',
                        help='URL of the CiviCRM module on the destination site.',
                        default='http://crm.example.org/sites/all/modules/civicrm')
    parser.add_argument('--sitekey',
                        help='The CiviCRM site_key of the site you are connecting to.',
                        default='')
    parser.add_argument('--apikey',
                        help='The CiviCRM api key.',
                        default='')
    parser.add_argument('--name',
                        help='Name of new mail template. Overrides a name specified by file.')
    parser.add_argument('--subject',
                        help='Email subject text. Overrides a subject specified by file.')
    parser.add_argument('--from_id',
                        help='CiviCRM Contact ID of sender.')
    inputgroup = parser.add_mutually_exclusive_group(required=True)
    inputgroup.add_argument('--json', nargs='?',
                            type=argparse.FileType('r'),
                            dest='jsonfile',
                            help='File containing the templated email as JSON.')
    inputgroup.add_argument('--url', nargs='?',
                            dest='jsonurl',
                            help='URL from which to fetch the templated email as JSON.')
    inputgroup.add_argument('--html', nargs='?',
                            type=argparse.FileType('r'),
                            dest='htmlfile',
                            help='File containing the templated HTML to email.')
    parser.add_argument('--text',
                        type=argparse.FileType('r'),
                        dest='textfile',
                        help='File containing the templated Text to email. If'
                             ' not supplied, the html version is rendered using'
                             ' html2text.')
    args = parser.parse_args()

    return args


def readjson(settings, jsontemplate):
    """
    Read the necesary input data for the mail template into the
    result, where metadata from the command line (seen in the
    global settings) overrides json-supplied data.

    :param jsontemplate: The JSON-derived input.
    :return: an object containing the data to send
    """
    result = CiviMailTemplate()
    result.name = jsontemplate['name']
    result.subject = jsontemplate['subject']
    result.from_id = jsontemplate['from_id']
    result.html = jsontemplate['html']
    if jsontemplate['plain']:
        result.plain = jsontemplate['plain']
    else:
        result.plain = getplaintext(result.html)

    if settings.name > "":
        result.name = settings.name
    if settings.subject > "":
        result.subject = settings.subject
    if settings.from_id > "":
        result.from_id = settings.from_id

    return result


def readlocal(settings):
    """
    Read the necesary input data for the mail template into the
    result, where metadata from the command line (seen in the
    global settings) overrides json-supplied data.

    :param jsontemplate: The JSON-derived input.
    :return: an object containing the data to send
    """
    result = CiviMailTemplate()
    result.name = settings.name
    result.subject = settings.subject
    result.from_id = settings.from_id
    result.html = settings.htmlfile.read()
    if settings.textfile:
        result.plain = settings.textfile.read()
    else:
        result.plain = getplaintext(result.html)

    return result


def fetch_url(jsonurl):
    """
    Read a JSON template for the email by fetching the URL provided.

    :param jsonurl: A URL that resolves to return application/json data.
    :return: The returned JSON content, or the null-json '{}'
    """
    r = requests.get(jsonurl)
    if r.status_code == 200 and r.headers['content-type'].startswith('application/json'):
        # Should be this instead? return r.text.encode('utf8')
        return r.text
    else:
        return '{}'


def getplaintext(html):
    """
    Return a reasonable plain-text version of the HTML input.

    :param html: string containing HTML input text.
    :return: string containing plain-text equivalent.
    """
    return html2text.html2text(html)


def connect_to_civi(settings):
    """
    Create a new CiviCRM object from the values in 'settings'.
    :return:
    """
    civicrm = CiviCRM(settings.civicrm, site_key=settings.sitekey,
                      api_key=settings.apikey, use_ssl=False)
    return civicrm


def check_creator_exists(settings, civicrm, creator_id):
    """
    Check that creator_id is a valid CiviCRM user.

    :param civicrm: Object used to talk to the CiviCRM api.
    :param creator_id: A Civicrm userid.
    :return: Boolean - True if the userid exists in CiviCRM as a user.
    """
    params = {
        u'contact_id': creator_id,
    }
    contactresults = civicrm.get(u'Contact', **params)
    debugmsg(settings, u'Owner is object ', contactresults)
    if len(contactresults) == 1:
        if contactresults[0][u'contact_id'] == creator_id:
            infomsg(settings, u'Creator is ', contactresults[0][u'sort_name'])
            return True
        else:
            warningmsg(settings, u'Creator id did not match : ' +
                       contactresults[0][u'contact_id'] + u' <> ' + creator_id)
            return False
    else:
        warningmsg(settings, u'Creator id was not found in CiviCRM.')
        return False


def create_template(settings, civicrm, template):
    """
    Send the email defined by the template to the CiviCRM instance.

    :param civicrm: Object defining a CiviCRM instance.
    :param template: Array defining the template mail.
    """
    params = {
        u'name': template.name,
        u'subject': template.subject,
        u'created_id': template.from_id,
        u'body_html': template.html,
        u'body_text': template.plain,
        u'url_tracking': u'1',
    }
    try:
        results = civicrm.create(u'Mailing', **params)
        debugmsg(settings, u'Returned object ', results)
        infomsg(settings, u'Created on:', results[u'created_date'])
        return True

    except CivicrmError as e:
        print(u'Mail template creation failed: ' + e.message)
        return False


def main():
    """
    Parse the command line args to determine where the mail template
    is coming from, fetch it, and send it on to the CiviCRM instance
    using the supplied URL and keys.

    Returns the integer code to shell:
        0 for success,
        1 for parameter problem,
        2 for internal error.
    """
    settings = getoptions()

    # There is a hierarchy of input sources: local HTML files are preferred,
    # then a local JSON file, then a JSON URL. However this should not
    # be important because 'getoptions()' considers the three sources to be
    # mutually exclusive.
    if settings.htmlfile:
        template = readlocal(settings)
    elif settings.jsonfile:
        jsontemplate = json.load(settings, settings.jsonfile)
        template = readjson(settings, jsontemplate)
    elif settings.jsonurl:
        jsontemplate = json.loads(fetch_url(settings.jsonurl))
        template = readjson(settings, jsontemplate)

    infomsg(settings, 'Using URL :', settings.civicrm)
    infomsg(settings, 'Name      :', template.name)
    infomsg(settings, 'Subject   :', template.subject)
    infomsg(settings, 'Creator   :', template.from_id)

    civicrm = connect_to_civi(settings)
    if check_creator_exists(settings, civicrm, template.from_id):
        if create_template(settings, civicrm, template):
            return 0

    return 1


try:
    if __name__ == "__main__":
        sys.exit(main())
    else:
        pass
except KeyboardInterrupt:
    print("Interrupted\n")
    sys.exit(1)
